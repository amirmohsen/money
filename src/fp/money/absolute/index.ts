import { Money } from '../..';
import { convertBigNumberToStringInteger } from '../_internals';

/**
 * Return the absolute monetary value of the current value,
 * i.e., remove the minus sign if the value is below zero
 * @returns - new Money instance with the absolute value
 */
const absolute = (money: Money): Money => {
  const newValue = money.value.absoluteValue();
  return new Money(convertBigNumberToStringInteger(money, newValue), money.currency);
};

export default absolute;
