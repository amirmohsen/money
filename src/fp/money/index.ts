export { default as money } from './money';
export { default as add } from './add';
export { default as subtract } from './subtract';
export { default as multiply } from './multiply';
export { default as divide } from './divide';
export { default as equals } from './equals';
export { default as clone } from './clone';
export { default as smallestUnit } from './smallestUnit';
