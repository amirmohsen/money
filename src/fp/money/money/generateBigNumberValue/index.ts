import BigNumber from 'bignumber.js';

const generateBigNumberValue = (BN: typeof BigNumber) => (value: string): BigNumber => new BN(value);

export default generateBigNumberValue;
