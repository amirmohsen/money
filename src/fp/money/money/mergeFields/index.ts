import BigNumber from 'bignumber.js';
import { MoneySymbol } from 'src/symbols';
import { BaseMoney, FrozenBaseCurrency } from '../../../types';

const mergeFields = (currency: FrozenBaseCurrency) => (value: BigNumber): BaseMoney =>
  Object.assign(Object.create(null), {
    value,
    currency,
    $$typeof: MoneySymbol,
  });

export default mergeFields;
