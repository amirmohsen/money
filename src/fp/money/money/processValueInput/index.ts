const processValueInput = (value: number | string | Money, BN: typeof BigNumber): BigNumber => {
  if (value instanceof Money) {
    this.checkValueCurrency(value);
    return value.amountAsBigNumber;
  }

  const divisor = getSmallestUnitDivisor(this);

  if (divisor.isGreaterThan(1) && ((typeof value === 'string' && isInt(value)) || Number.isInteger(value as number))) {
    const bignumber = new BN(value);

    return bignumber.dividedBy(divisor).decimalPlaces(this.currency.decimalDigits);
  }

  if (typeof value === 'string' && isFloat(value)) {
    return new BN(value);
  }

  throw new WrongInputError(
    oneLine`
    The input value must be either an integer,
    an integer-like string, a float-like string or a "Money" instance.
    `,
  );
};

export default processValueInput;
