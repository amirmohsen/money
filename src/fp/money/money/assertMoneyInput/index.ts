import { WrongInputError } from 'src/shared/errors';
import { FrozenBaseCurrency } from 'src/fp/types';

const assertInt = (value: string): void => {
  if (!/^[-|+]?\d+$/.exec(value)) {
    throw new WrongInputError(
      `The input value, "${value}", is not an integer to match the currency's 0 decimal digits.`,
    );
  }
};

const assertFixedDigitDecimal = (value: string, decimalDigits: number): void => {
  if (!new RegExp(`^[-|+]?\\d+\\.\\d{${decimalDigits}}$`).exec(value)) {
    throw new WrongInputError(
      `The input value, "${value}", is not a fixed decimal number to match the currency's ${decimalDigits} decimal digits.`,
    );
  }
};

const assertValidNumber = (value: string, decimalDigits: number): void =>
  decimalDigits === 0 ? assertInt(value) : assertFixedDigitDecimal(value, decimalDigits);

const assertMoneyInput = (value: string, currency: FrozenBaseCurrency): string => {
  if (typeof value !== 'string') {
    throw new WrongInputError(`The input value, "${value}", is not a string.`);
  }
  assertValidNumber(value, currency.decimalDigits);
  return value;
};

export default assertMoneyInput;
