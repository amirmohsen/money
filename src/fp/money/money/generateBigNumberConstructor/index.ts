import BigNumber from 'bignumber.js';
import ROUNDING from 'src/shared/constants/ROUNDING';
import { FrozenBaseCurrency } from 'src/fp/types';

const generateBigNumberConstructor = ({
  decimalSeparator,
  thousandsSeparator: groupSeparator,
}: FrozenBaseCurrency): typeof BigNumber => {
  return BigNumber.clone({
    DECIMAL_PLACES: 20,
    ROUNDING_MODE: ROUNDING.HALF_UP as any,
    FORMAT: {
      decimalSeparator,
      groupSeparator,
      groupSize: 3,
      secondaryGroupSize: 0,
      fractionGroupSeparator: ' ',
      fractionGroupSize: 0,
    },
  });
};

export default generateBigNumberConstructor;
