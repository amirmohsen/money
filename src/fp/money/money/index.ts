import deepFreeze from 'deep-freeze';
import pipe from 'lodash/fp/pipe';
import { FrozenBaseCurrency, FrozenBaseMoney } from 'src/fp/types';
import assertMoneyInput from './assertMoneyInput';
import generateBigNumberValue from './generateBigNumberValue';
import generateBigNumberConstructor from './generateBigNumberConstructor';
import mergeFields from './mergeFields';

const money = (value: string, currency: FrozenBaseCurrency): FrozenBaseMoney =>
  pipe(
    assertMoneyInput,
    pipe(
      generateBigNumberConstructor,
      generateBigNumberValue,
    )(currency),
    mergeFields(currency),
    deepFreeze,
  )(value, currency) as FrozenBaseMoney;

export default money;
