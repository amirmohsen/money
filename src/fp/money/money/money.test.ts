/* eslint-disable jest/no-identical-title */
import { currency } from 'src/fp/currency';
import { WrongInputError } from 'src/shared/errors';
import { FrozenBaseCurrency, FrozenBaseMoney } from 'src/fp/types';
import BigNumber from 'bignumber.js';
import money from '.';

describe('money', () => {
  const doubleDigitCurrency = currency('USD');
  const doubleDigitCurrencyWithDifferentFormatting = currency('EUR', {
    thousandsSeparator: '.',
    decimalSeparator: ',',
  });
  const zeroDigitCurrency = currency('JPY', { decimalDigits: 0 });

  describe.each([[false], [true], [null], [undefined], [34], [234.3454], [-23423], [{}], [[]]] as any[][])(
    'when given any non-string value',
    (value: any) => {
      test('should throw an error', () => {
        expect(() => money(value, doubleDigitCurrency)).toThrow(
          new WrongInputError(`The input value, "${value}", is not a string.`),
        );
      });
    },
  );

  describe.each([[''], ['test'], ['123.45'], ['-34.5666778'], ['abc123']])(
    'when given a non-integer string value with a zero-decimal-digit currency',
    (value: string) => {
      test('should throw an error', () => {
        expect(() => money(value, zeroDigitCurrency)).toThrow(
          new WrongInputError(
            `The input value, "${value}", is not an integer to match the currency's 0 decimal digits.`,
          ),
        );
      });
    },
  );

  describe.each([[''], ['test'], ['123'], ['-345666778'], ['abc123'], ['34.3456'], ['346576.2', '34.0']])(
    "when given a non-decimal string value or a decimal value that doesnt't match the currency's decimal digit count",
    (value: string) => {
      test('should throw an error', () => {
        expect(() => money(value, doubleDigitCurrency)).toThrow(
          new WrongInputError(
            `The input value, "${value}", is not a fixed decimal number to match the currency's 2 decimal digits.`,
          ),
        );
      });
    },
  );

  describe.each(([
    ['123334', zeroDigitCurrency],
    ['1234232.90', doubleDigitCurrency],
    ['34233.89', doubleDigitCurrencyWithDifferentFormatting],
  ] as unknown) as [string, FrozenBaseCurrency][])(
    'when provided with valid values',
    (value: string, cur: FrozenBaseCurrency) => {
      let moneyObj: FrozenBaseMoney;

      beforeAll(() => {
        moneyObj = money(value, cur);
      });

      test('should create a Money object', () => {
        expect(moneyObj).toMatchSnapshot();
      });

      test('should use a BigNumber object as the value', () => {
        expect(BigNumber.isBigNumber(moneyObj.value)).toBe(true);
      });

      test('should create a frozen object', () => {
        expect(Object.isFrozen(moneyObj)).toBe(true);
      });
    },
  );
});
