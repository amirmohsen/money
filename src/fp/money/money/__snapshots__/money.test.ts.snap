// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`money when provided with valid values should create a Money object 1`] = `
Object {
  "$$typeof": Symbol(wealth.money),
  "currency": Object {
    "$$typeof": Symbol(wealth.currency),
    "code": "JPY",
    "decimalDigits": 0,
    "decimalSeparator": ".",
    "pattern": "%ns%s%v",
    "symbol": "JPY",
    "thousandsSeparator": ",",
  },
  "value": "123334",
}
`;

exports[`money when provided with valid values should create a Money object 2`] = `
Object {
  "$$typeof": Symbol(wealth.money),
  "currency": Object {
    "$$typeof": Symbol(wealth.currency),
    "code": "USD",
    "decimalDigits": 2,
    "decimalSeparator": ".",
    "pattern": "%ns%s%v",
    "symbol": "USD",
    "thousandsSeparator": ",",
  },
  "value": "1234232.9",
}
`;

exports[`money when provided with valid values should create a Money object 3`] = `
Object {
  "$$typeof": Symbol(wealth.money),
  "currency": Object {
    "$$typeof": Symbol(wealth.currency),
    "code": "EUR",
    "decimalDigits": 2,
    "decimalSeparator": ",",
    "pattern": "%ns%s%v",
    "symbol": "EUR",
    "thousandsSeparator": ".",
  },
  "value": "34233.89",
}
`;
